x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java

```diff
@@ -754,89 +681,133 @@ protected void alreadyClosed() {
+                // find out what gaps we must fill for the given region
+                final ByteRange subRangeToRead = mapSubRangeToRegion(rangeToRead, region);
+                final ByteRange subRangeToWrite = mapSubRangeToRegion(rangeToWrite, region);
+                assert subRangeToRead.length() > 0 || subRangeToWrite.length() > 0
+                    : "Either read or write region must be non-empty but saw [" + subRangeToRead + "][" + subRangeToWrite + "]";
+
+                final List<SparseFileTracker.Gap> gaps = fileRegion.tracker.waitForRange(
                     subRangeToWrite,
                     subRangeToRead,
-                    (channel, channelPos, relativePos, length) -> {
-                        assert regionOwners[fileRegion.sharedBytesPos].get() == fileRegion;
-                        assert channelPos >= fileRegion.physicalStartOffset() && channelPos + length <= fileRegion.physicalEndOffset();
-                        return reader.onRangeAvailable(channel, channelPos, relativePos - readOffset, length);
-                    },
-                    (channel, channelPos, relativePos, length, progressUpdater) -> {
-                        assert regionOwners[fileRegion.sharedBytesPos].get() == fileRegion;
-                        assert channelPos >= fileRegion.physicalStartOffset() && channelPos + length <= fileRegion.physicalEndOffset();
-                        writer.fillCacheRange(channel, channelPos, relativePos - writeOffset, length, progressUpdater);
-                    },
-                    executor
+                    fileRegion.rangeListener(subRangeToRead, readRangeRelativeStart, reader, listener, fileChannel)
                 );
-                assert lis != null;
+
+                if (gaps.isEmpty() == false) {
+                    context.addRegionToPopulate(fileChannel, fileRegion, gaps, regionStart);
+                } else {
+                    context.skipRegion(fileChannel, fileRegion);
+                }
+
                 if (stepListener == null) {
-                    stepListener = lis;
+                    stepListener = listener;
                 } else {
-                    stepListener = stepListener.thenCombine(lis, Math::addExact);
+                    stepListener = stepListener.thenCombine(listener, Math::addExact);
                 }
+            }
 
+            // If there are gaps to fill do so via the provided executor, if not then the stepListener is already resolved
+            if (context.regionsToIterate() > 0) {
+                final ByteBufferConsumer handler = new ByteBufferConsumer(context);
+                final long relativeWritePos = context.writeStartAbsolute - rangeToWrite.start();
+                executor.execute(new AbstractRunnable() {
+                    @Override
+                    protected void doRun() throws Exception {
+                        writer.fillCacheRange(handler, relativeWritePos, context.bytesToRead());
```

Comment:
I found it confusing to mix `handler` and `context` here, so maybe better to declare the `ByteBufferConsumer` only where it is used (in `doRun()`) and use context elsewhere?